eureka-core/src/main/java/com/netflix/eureka2/interests/InstanceInfoInitStateHolder.java

```diff
@@ -1,40 +1,76 @@
 package com.netflix.eureka2.interests;
 
-import com.netflix.eureka2.registry.instance.InstanceInfo;
-
 import java.util.Iterator;
-import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ConcurrentHashMap;
+
+import com.netflix.eureka2.interests.ChangeNotification.Kind;
+import com.netflix.eureka2.registry.instance.InstanceInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
- * An {@link Index.InitStateHolder} implementation for {@link InstanceInfo}
+ * An {@link Index.InitStateHolder} implementation for {@link InstanceInfo}.
+ * As the cached state is backed by {@link ConcurrentHashMap}, the order of notifications
+ * is not preserved, which is fine as there is exactly one item per {@link InstanceInfo} object.
  *
  * @author Nitesh Kant
  */
 public class InstanceInfoInitStateHolder extends Index.InitStateHolder<InstanceInfo> {
 
-    // TODO: Implement compaction
-    private final ConcurrentLinkedQueue<ChangeNotification<InstanceInfo>> notifications;
+    private static final Logger logger = LoggerFactory.getLogger(InstanceInfoInitStateHolder.class);
+
+     final ConcurrentHashMap<String, ChangeNotification<InstanceInfo>> notificationMap;
 
     public InstanceInfoInitStateHolder(Iterator<ChangeNotification<InstanceInfo>> initialRegistry) {
         super(NotificationsSubject.<InstanceInfo>create());
-        notifications = new ConcurrentLinkedQueue<ChangeNotification<InstanceInfo>>();
+        notificationMap = new ConcurrentHashMap<>();
+
         while (initialRegistry.hasNext()) {
-            notifications.add(initialRegistry.next());
+            ChangeNotification<InstanceInfo> next = initialRegistry.next();
+            notificationMap.put(next.getData().getId(), next); // Always Kind.Add
         }
     }
 
     @Override
     protected void addNotification(ChangeNotification<InstanceInfo> notification) {
-        notifications.add(notification);
+        String id = notification.getData().getId();
+        ChangeNotification<InstanceInfo> current = notificationMap.get(id);
+
+        ChangeNotification<InstanceInfo> updated = processNext(current, notification);
```

Comment:
Updates are serialized and come from the registry. Also for concurrent updates, it is thread safe. The issue might be only when the same instance info is updated concurrently from two different threads. In such case we would end up with the problem we have in the registry. But fortunately this is not the case.